A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,668 @@
+  map<string, Action> action = 2;
+}
+```
+
+The config must fulfill the following:
+
+*   Routes
+    * Matchers
+        *   There must be exactly one `path_matcher`.
+            *   `path_matcher` is a oneof so this is always true in proto. But
+                json doesn’t have the oneof concept, so a route may have
+                multiple path matchers, and it will be considered invalid.
+        *   There can be one or more optional `header matchers`.
+        *   There can be one optional `match_fraction`.
+    *   The `action` name must be present in the action map.
+    *   Duplicate matchers are allowed.
+        *   Duplicate matcher means that more than one routes have the same
+            matcher, no matter whether the actions are the same or not.
+        *   Because the first match wins, the second matcher will never be used.
+*   Actions
+    *   Every action in the map must be referred to by at least one route.
+    *   There must be at least one valid config in `child_policy`.
+
+Violating any of these will make the LB policy config invalid, and be rejected.
+
+For each pick, a route is matched if all matching criteria are matched. The
+first match wins.
+
+For each action, the config in `child_policy` to be used to create a child LB
+policy will be picked using the same algorithm as [service
+config](https://github.com/grpc/grpc-proto/blob/dd2dca318eb197b96b60f22297871fb1ed862800/grpc/service_config/service_config.proto#L401).
+
+*   The first known (registered) policy name in the list will be picked
+    *   If none is registered, the config is invalid
+*   The config for the picked policy will be validated
+    *   If the validation fails, the config is invalid. We won’t move on to the
+        next entry
+
+Action names are used as ID for the child policies. If an action name exists in
+the previous service config, the action’s LB policy config will be sent as an
+update to the corresponding child policy (instead of creating a new child policy
+for it). The route refers to the action name instead of embedding the action
+directly, so that one action can be shared by multiple routes.
+
+NOTE: this LB policy is named `xds_routing` (instead of just routing) because we
+plan to make it deeply integrated with the xds resolver as part of the
+forthcoming RouteAction design.
+
+### Picking
+
+When making a pick, the first match wins.
+
+For example, with config
+
+```
+{prefix: "/MyService"}:             childPolicy-1
+{path:   "/MyService/MyMethod"}:    childPolicy-2
+```
+
+`ChildPolicy-1` will be picked for `/MyService/MyMethod`, even though the
+second route is an exact match. It is this way because Envoy always picks the
+first matched route. And with header matching, finding the exact match can be
+tricky.
+
+The pick will always be delegated to the picker from the matched child
+policy, even if it’s not in connectivity state READY (as opposed to
+weighted\_target, which will only pick the READY child policies).
```

Comment:
Why does this distinction exist in gRPC? I don't think this happens in Envoy.

Reply:
In gRPC, it's normal for an LB policy to monitor the connectivity state of its children and not send traffic to children that are not reporting READY.  I understand that in the WeightedClusters case, Envoy does not do this, but I think that behavior would be surprising to gRPC users.  And it seems like the Envoy behavior is kind of an accident based on where the WeightedClusters functionality was implemented, rather than something that users would actually want; if we instead moved that functionality to an LB policy in Envoy (as per discussion in https://github.com/envoyproxy/envoy/issues/5598), then it probably would have the same behavior as gRPC.

Reply:
Fair, but I think it would help to provide some wording on the context here.

Reply:
Good point.  Menghan, please document this.

Reply:
What kind of explanation do you have in mind?

The weighted_target behavior is the normal gRPC balancing policy behavior, the routing policy is the special one, so there's this one sentence explaining that.

Do you want me to talk about the difference between Envoy and gRPC here?
Or one more sentence on why routing is different from weighted_target?

Reply:
I think it will be useful to point out both (a) differences from Envoy behavior for those who have experience with Envoy but not gRPC and (b) differences from normal gRPC behavior for those that have experience with gRPC and are now starting to use xDS.

Specifically:
- In the xds_routing section, we should mention that, unlike most gRPC LB policies, the xds_routing policy will not ignore children that are not reporting READY.
- In the weighted_target section, we should mention that, unlike Envoy, gRPC will not send any traffic to clusters in the WeightedClusters list that are not reporting READY.

Reply:
Done. Added explanation for weighted_target and routing policy

Reply:
Looks great (modulo one minor wording nit, which I pointed out below).